### PR TITLE
Fix shadowing agent causing episode done

### LINKED ISSUE #58: Shadowing actor should not trigger done
Subtle bug w/ airlocking: if the shadowing actor returns done, it causes the vehicle to terminate. This should not be possible. Instead, only when the actor is actually in control can it influence vehicle termination. The solution likely involves updating the agent manager's observe method.

Add tests for removal in airlock and bubble.

### PR DESCRIPTION
**Pull request by JVillella**
closes #58

### CODE REVIEW

File: smarts/core/smarts.py [hunk 1]
```diff
@@ -255,39 +255,40 @@ def _step(self, agent_actions):
...
-                        if done[vehicle_id]:
-                            vehicle_ids.add(vehicle_id)
+                    vehicle_ids.update([id_ for id_ in done if done[id_]])
                 elif done:
-                    ids = vehicle_index.vehicle_ids_by_actor_id(agent_id)
-                    assert len(ids) == 1
-                    vehicle_ids.add(ids[0])
+                    ids = self._vehicle_index.vehicle_ids_by_actor_id(agent_id)
+                    # 0 if shadowing, 1 if active
+                    assert len(ids) <= 1, f"{len(ids)} <= 1"
```
Comment (aurorahcx) [inline smarts/core/smarts.py:266]: I might be missing something, what's the change that makes len(ids) = 0 possible here comparing to before?

File: smarts/core/smarts.py [hunk 2]
```diff
@@ -255,39 +255,40 @@ def _step(self, agent_actions):
...
+
+        # XXX: These can not be put inline because we do queries that must proceed
+        #      the actual teardown.
+        vehicles_to_teardown = done_vehicle_ids(dones)
+        agents_to_teardown = done_agent_ids(dones)
 
-        self._teardown_agent_vehicles(done_vehicle_ids(dones, self._vehicle_index))
+        self._agent_manager.teardown_ego_agents(agents_to_teardown)
+        self._agent_manager.teardown_social_agents(agents_to_teardown)
+        self._teardown_agent_vehicles(vehicles_to_teardown)
```
Comment (aurorahcx) [inline smarts/core/smarts.py:291]: _teardown_agent_vehicles() takes agent_ids instead of vehicle_ids as as argument.

Comment (JVillella) [reply smarts/core/smarts.py:291]: not anymore :)

Comment (aurorahcx) [reply smarts/core/smarts.py:291]: Sorry I was looking at the master branch.

File: smarts/core/smarts.py [hunk 3]
```diff
@@ -255,39 +255,40 @@ def _step(self, agent_actions):
...
-        def done_vehicle_ids(dones, vehicle_index):
+        def done_vehicle_ids(dones):
             vehicle_ids = set()
             for agent_id, done in dones.items():
                 if self._agent_manager.is_boid_agent(agent_id):
-                    for vehicle_id in done:
-                        if done[vehicle_id]:
-                            vehicle_ids.add(vehicle_id)
+                    vehicle_ids.update([id_ for id_ in done if done[id_]])
                 elif done:
```
Comment (Gamenot) [inline smarts/core/smarts.py:263]: Looking at this, it seems awkward to me that these are handled so differently.

File: smarts/core/smarts.py [hunk 4]
```diff
@@ -255,39 +255,40 @@ def _step(self, agent_actions):
...
-            agents_to_teardown
-        )
+        def done_agent_ids(dones):
+            agent_ids = set()
+            for agent_id, done in dones.items():
+                if self._agent_manager.is_boid_agent(agent_id):
+                    if not self.agent_manager.is_boid_keep_alive_agent(
+                        agent_id
+                    ) and all(dones[agent_id].values()):
+                        agent_ids.add(agent_id)
```
Comment (Gamenot) [inline smarts/core/smarts.py:278]: Double if.

```suggestion
                if (
                    self._agent_manager.is_boid_agent(agent_id)
                    and not self.agent_manager.is_boid_keep_alive_agent(agent_id)
                    and all(dones[agent_id].values())
                ):
                    agent_ids.add(agent_id)
```

Comment (JVillella) [reply smarts/core/smarts.py:278]: Intentional! Notice where the else clause is.

File: smarts/core/smarts.py (hunk 3 above)
Comment (JVillella) [reply smarts/core/smarts.py:263]: It's due to boids having a second level dictionary (of vehicles).

**Review by aurorahcx — APPROVED** [review]

**Review by liamchzh — APPROVED** [review]

**Review by davidrusu — APPROVED** [review]
Just a nit, besides that looks great

File: smarts/core/smarts.py [hunk 5]
```diff
@@ -255,39 +255,40 @@ def _step(self, agent_actions):
...
-
-        def done_vehicle_ids(dones, vehicle_index):
+        def done_vehicle_ids(dones):
             vehicle_ids = set()
             for agent_id, done in dones.items():
                 if self._agent_manager.is_boid_agent(agent_id):
-                    for vehicle_id in done:
-                        if done[vehicle_id]:
-                            vehicle_ids.add(vehicle_id)
+                    vehicle_ids.update([id_ for id_ in done if done[id_]])
```
[the remark this answers is not in the document]
Comment (JVillella) [reply smarts/core/smarts.py:262]: oh great tip!

**Pull request merged**